path/path.go

```diff
@@ -3,187 +3,240 @@ package path
+	IPNSNamespace
+	IPLDNamespace
+)
+
+// Path is a generic, valid, and well-formed path. A valid path is shaped as follows:
+//
+//	/{namespace}/{root}[/remaining/path]
+//
+// Where:
+//
+//  1. Namespace is "ipfs", "ipld", or "ipns".
+//  2. If namespace is "ipfs" or "ipld", "root" must be a valid [cid.Cid].
+//  3. If namespace is "ipns", "root" may be a [ipns.Name] or a DNSLink FQDN.
+//
+// [DNSLink]: https://dnslink.dev/
+type Path interface {
+	// String returns the path as a string.
+	String() string
+
+	// Namespace returns the first component of the path. For example, the namespace
+	// of "/ipfs/bafy" is "ipfs".
+	Namespace() Namespace
+
+	// Segments returns the different elements of a path delimited by a forward
+	// slash ("/"). The returned array must not contain any empty segments, and
+	// must have a length of at least two: the first element must be the namespace,
+	// and the second must be root.
+	//
+	// Examples:
+	// 		- "/ipld/bafkqaaa" returns ["ipld", "bafkqaaa"]
+	// 		- "/ipfs/bafkqaaa/a/b/" returns ["ipfs", "bafkqaaa", "a", "b"]
+	// 		- "/ipns/dnslink.net" returns ["ipns", "dnslink.net"]
+	Segments() []string
 }
 
-// Segments returns the different elements of a path
-// (elements are delimited by a /).
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+var _ Path = path{}
 
-	// Ignore leading slash
-	if len(segments[0]) == 0 {
-		segments = segments[1:]
-	}
+type path struct {
+	str       string
+	namespace Namespace
+}
 
-	return segments
+func (p path) String() string {
+	return p.str
 }
 
-// String converts a path to string.
-func (p Path) String() string {
-	return string(p)
+func (p path) Namespace() Namespace {
+	return p.namespace
 }
 
-// IsJustAKey returns true if the path is of the form <key> or /ipfs/<key>, or
-// /ipld/<key>
-func (p Path) IsJustAKey() bool {
-	parts := p.Segments()
-	return len(parts) == 2 && (parts[0] == "ipfs" || parts[0] == "ipld")
+func (p path) Segments() []string {
+	// Trim slashes from beginning and end, such that we do not return empty segments.
+	str := strings.TrimSuffix(p.str, "/")
+	str = strings.TrimPrefix(str, "/")
+
+	return strings.Split(str, "/")
 }
 
-// PopLastSegment returns a new Path without its final segment, and the final
-// segment, separately. If there is no more to pop (the path is just a key),
-// the original path is returned.
-func (p Path) PopLastSegment() (Path, string, error) {
-	if p.IsJustAKey() {
-		return p, "", nil
-	}
+// ImmutablePath is a [Path] which is guaranteed to have an immutable [Namespace].
+type ImmutablePath interface {
+	Path
 
-	segs := p.Segments()
-	newPath, err := ParsePath("/" + strings.Join(segs[:len(segs)-1], "/"))
-	if err != nil {
-		return "", "", err
-	}
+	// Cid returns the [cid.Cid] of the root object of the path.
+	Cid() cid.Cid
 
-	return newPath, segs[len(segs)-1], nil
+	// Remainder returns the unresolved parts of the path.
+	Remainder() string
```

Comment:
> is what's left to be resolved, or cannot be resolved

But this indicates that it's just any path components beyond the CID, doesn't it? https://github.com/ipfs/boxo/pull/334/files#diff-a44166f2b1bbfa22a31664836b3206fbcfb31600bac5a85b0837aa0644131951R142-R147

> I don't understand this question. We're not returning a []string.

We're returning `[]string`  in places like `resolver.ResolveToLastNode` https://github.com/ipfs/boxo/pull/334/files#diff-0731f9658c3343bd65daf65521d04f5064c46ac26dd0d9bce182db016a8f0ba4R82. That's not a change in this PR, but it's somewhere that returning a path with a `Remainder` might be plausible depending on how we've defined remainder.

---

An area where I'm not sure if there are ramifications here, but maybe you can help me understand since you've been going through all of this now, is if it matters now that we:
- Used to have: ResolvedPath has Root CID, Path, Terminal CID, remainder
- Now have: ImmutablePath has Root CID, Path (remainder == path)

This loss of remainder information seems like it could be problematic. For example, f it was meant to convey "hey I tried resolving through this block but there are no more links, just other IPLD data model fields" then that information would need to be recomputed or specified as required in other interface contracts.

Reply:
@aschmahmann there are surprisingly little places where we use the concept of `.Remainder()` in both Boxo and Kubo. Boxo mainly uses it in the [handler codec](https://github.com/ipfs/boxo/blob/32140848b581e44bc044a66af149814732174915/gateway/handler_codec.go#L84-L92) to ensure there's nothing going inside the path, and Kubo only uses it as a result of [`ResolvePath`
](https://github.com/search?q=repo%3Aipfs%2Fkubo+Remainder%28%29&type=code). I have a suggestion: remove `Remainder` from `ImmutablePath`, and change the `ResolvePath` signature to return `ImmutablePath, []string, error`, matching `ResolveToLastNode`. What do you think? This way, we remove the ambiguous value in the path type and keep it for resolution.

Reply:
@aschmahmann I made the changes mentioned above:

- Boxo: https://github.com/ipfs/boxo/pull/334/commits/4276346cc566a4033b069d08af9d0ea7323119fc
- Kubo: https://github.com/ipfs/kubo/pull/10063/commits/f5b41755bfe4033167f20c6a787802849a14fed5

Let me know what you think. I think this makes the APIs more consistent, and removes the unclear `.Remainder`. Now the remainder is only returned by resolvers and resolving functions, making it clearer what it means and what it should be used for.